### PR TITLE
Add Spotify support

### PR DESCRIPTION
**Pull request by tkozakas**

### CODE REVIEW

File: handlers.ts [hunk 1]
```diff
@@ -26,12 +26,88 @@ function isYoutubeUrl(url: string): boolean {
...
+
 async function getVideoTitle(url: string): Promise<string> {
-  if (isYoutubeUrl(url) && process.env.YTDLP_USE_OAUTH_PLUGIN === "true") {
-    return $`yt-dlp --username oauth2 --password unused --get-title -- "${url}"`.text();
-  } else {
-    return $`yt-dlp --get-title -- "${url}"`.text();
-  }
+    const youtubeUrl = await handleUrl(url);
+    return $`yt-dlp --get-title -- "${youtubeUrl}"`.text();
 }
```
Comment (neiios) [inline handlers.ts:111]: this will break on the server

**Review by neiios — CHANGES_REQUESTED** [review]

File: handlers.ts [hunk 2]
```diff
@@ -26,12 +26,88 @@ function isYoutubeUrl(url: string): boolean {
...
+    if (!response.ok) {
+        throw new BotError("Failed to fetch track data from Spotify");
+    }
+
+    const data = await response.json();
+    const title = data.name;
+    const artist = data.artists[0].name;
+
+    const query = `${artist} ${title}`;
+    const searchResult = await $`yt-dlp "ytsearch:${query}" --get-id --default-search "ytsearch"`.text();
```
Comment (neiios) [inline handlers.ts:88]: should break the same way on the server

File: handlers.ts [hunk 3]
```diff
@@ -74,13 +140,13 @@ export async function handlePlay(
...
 
   if (PLAYER.state.status === "playing" || LOCKED) {
-    const [videoTitle, downloadedAudio] = await Promise.all([
-      getVideoTitle(url),
-      downloadAudio(url),
-    ]);
+      const [videoTitle, downloadedAudio] = await Promise.all([
+          getVideoTitle(url),
+          downloadAudio(url),
+      ]);
```
Comment (neiios) [inline handlers.ts:149]: prettier sir?

File: handlers.ts [hunk 4]
```diff
@@ -26,12 +26,88 @@ function isYoutubeUrl(url: string): boolean {
...
+async function handleUrl(url: string): Promise<string> {
+    if (isSpotifyUrl(url)) {
+        return await fetchYouTubeUrlFromSpotify(url);
+    } else if (isYoutubeUrl(url)) {
+        return url;
+    } else {
+        throw new BotError("Unsupported URL format");
+    }
+}
+
```
Comment (neiios) [inline handlers.ts:107]: also this will fail for other urls other than youtube or spotify

**Review by neiios — CHANGES_REQUESTED** [review]

File: handlers.ts (hunk 4 above)
Comment (neiios) [reply handlers.ts:107]: the intended behavior was to use the default yt-dl command syntax on other websites

File: handlers.ts (hunk 1 above)
Comment (tkozakas) [reply handlers.ts]: Reverted

File: handlers.ts (hunk 4 above)
Comment (tkozakas) [reply handlers.ts]: Fixed here: 9da044821a87264494f0aa97bc91ad24d8e088b4

File: handlers.ts (hunk 2 above)
Comment (tkozakas) [reply handlers.ts]: Should be fixed: 9da044821a87264494f0aa97bc91ad24d8e088b4

**Review by neiios — APPROVED** [review]
YOLO LGTM

**Pull request merged**